[BUG] Inserting into emoticon to "char(1)" type column raise exception

**Describe the bug**
<!-- A clear and concise description of what the bug is. -->
Inserting into emoticon to "char(1)" type column raise exception
**What did you do**
<!-- 
If possible, please provide a code receipt to produce this issue.
e.g., 
1. run `spark-shell --jars xxx.jar`
2. in spark-shell, run 
```
val df = spark.sql(“select sum(b) from t where a = 1”)
df.show(20)
```
3. update some data through TiDB
4. in spark-shell, run
```
val df = spark.sql(“select sum(b) from t where a = 1”)
df.show(20)
```
 -->
1. execute in tidb
use test;
CREATE TABLE t1 (c char(1) NOT NULL)
INSERT INTO t1 VALUES('😉')
2. execute in tispark
create table st using tidb options(table 't1')
INSERT INTO st1 VALUES ('😉')
INSERT INTO st1 select * from test.t1

**What do you expect**
<!-- A clear and concise description of what you expected to happen. -->
successfully
**What happens instead**
<!-- If an error occurs, please provide complete error stack. -->

<!-- 
**Screenshots**
If applicable, add screenshots to help explain your problem.
 -->
```
Caused by: com.pingcap.tikv.exception.TiDBConvertException: convert to tidb data error for column 'c'
	at com.pingcap.tispark.write.TiBatchWriteTable$$anonfun$com$pingcap$tispark$write$TiBatchWriteTable$$sparkRow2TiKVRow$1.apply$mcVI$sp(TiBatchWriteTable.scala:565)
	... 30 more
Caused by: com.pingcap.tikv.exception.ConvertOverflowException: value 😉 length > max length 1
	at com.pingcap.tikv.exception.ConvertOverflowException.newMaxLengthException(ConvertOverflowException.java:28)
	at com.pingcap.tikv.types.StringType.convertToString(StringType.java:99)
	at com.pingcap.tikv.types.StringType.doConvertToTiDBType(StringType.java:50)
	at com.pingcap.tikv.types.DataType.convertToTiDBType(DataType.java:395)
	at com.pingcap.tispark.write.TiBatchWriteTable$$anonfun$com$pingcap$tispark$write$TiBatchWriteTable$$sparkRow2TiKVRow$1.apply$mcVI$sp(TiBatchWriteTable.scala:560)
	... 30 more
```
**Spark and TiSpark version info**
<!-- What version of Spark and TiSpark are you using? (Provide Spark version and run `spark.sql(“select ti_version()”).show(false)` in spark-shell) -->

<!--
**Additional context**
Add any other context about the problem here.
You may also provide TiDB version here if it is related to the issue.
-->
Git Commit Hash: 9687b70202584883fe5b88d39fb62d64867ebde2
Git Branch: release-2.3
UTC Build Time: 2020-10-16 02:40:15
Supported Spark Version: 2.3 2.4
Current Spark Version: 2.4.3
Current Spark Major Version: 2.4
TimeZone: Asia/Shanghai

TIDB: release-4.0

--- a/core/src/test/scala/com/pingcap/tispark/overflow/StringOverflowSuite.scala
+++ b/core/src/test/scala/com/pingcap/tispark/overflow/StringOverflowSuite.scala
@@ -30,6 +30,40 @@ import org.apache.spark.sql.types.{StructField, _}
  */
 class StringOverflowSuite extends BaseBatchWriteTest("test_data_type_string_overflow") {
 
+  test("Test UTF-8 like emoji not overflow") {
+
+    compareTiDBWriteWithJDBC {
+      case (writeFunc, _) =>
+        jdbcUpdate(s"drop table if exists $dbtable")
+        jdbcUpdate(s"create table $dbtable(i int, c1 CHAR(1))")
+        val row = Row(1, "\uD83D\uDE09")
+        val schema =
+          StructType(List(StructField("i", IntegerType), StructField("c1", StringType)))
+        writeFunc(List(row), schema, None)
+        compareTiDBSelectWithJDBC(Seq(row), schema)
+    }
+  }
+
+  test("Test UTF-8 like emoji overflow") {
+
+    jdbcUpdate(s"drop table if exists $dbtable")
+    jdbcUpdate(s"create table $dbtable(c1 CHAR(1))")
+    val row = Row("\uD83D\uDE09\uD83D\uDE09")
+    val schema =
+      StructType(List(StructField("c1", StringType)))
+
+    val jdbcErrorClass = classOf[java.sql.BatchUpdateException]
+    val tidbErrorClass = classOf[com.pingcap.tikv.exception.ConvertOverflowException]
+    val tidbErrorMsg = "value \uD83D\uDE09\uD83D\uDE09 length > max length 1"
+
+    compareTiDBWriteFailureWithJDBC(
+      List(row),
+      schema,
+      jdbcErrorClass,
+      tidbErrorClass,
+      tidbErrorMsg)
+  }
+
   test("Test CHAR Overflow") {
     testCharOverflow(false)
   }

--- a/tikv-client/src/main/java/com/pingcap/tikv/types/StringType.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/types/StringType.java
@@ -96,7 +96,7 @@ public class StringType extends BytesType {
     }
 
     // check utf-8 length of result string
-    if (result.codePointCount(0, result.length() - 1) > this.getLength()) {
+    if (result.codePointCount(0, result.length()) > this.getLength()) {
       throw ConvertOverflowException.newMaxLengthException(result, this.getLength());
     }
 

--- a/tikv-client/src/main/java/com/pingcap/tikv/types/StringType.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/types/StringType.java
@@ -95,7 +95,8 @@ public class StringType extends BytesType {
       throw new ConvertNotSupportException(value.getClass().getName(), this.getClass().getName());
     }
 
-    if (result.length() > this.getLength()) {
+    // check utf-8 length of result string
+    if (result.codePointCount(0, result.length() - 1) > this.getLength()) {
       throw ConvertOverflowException.newMaxLengthException(result, this.getLength());
     }
 

